lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+-- @treturn boolean index If the operation is successful
+local function matrix_add(matrix, object, row, col, row_span, col_span)
+    if row < 1 then return false end
+    if col < 1 then return false end
+    row_span = row_span > 0 and row_span or 1
+    col_span = col_span > 0 and col_span or 1
+    -- loop over indices
+    for i=1, (row + row_span - 1) do
+        if not matrix[i] then matrix[i] = {} end
+        if i >= row then
+            for j=1, (col + col_span -1) do
+                if j >= col then
+                    matrix[i][j] = object
+                end
+            end
+        end
+    end
+    return true
+end
+
+--- Remove a widget from the matrix
+-- the widget is set in coordinates `row`:`col`, spanning several cells
+-- given by `row_span` and `col_span`
+-- @param matrix Matrix containing the widgets
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @param number row_span The number of rows the widget spans (default to 1)
+-- @param number col_span The number of columns the widget spans (default to 1)
+-- @treturn boolean index If the operation is successful
+local function matrix_remove(matrix, row, col, row_span, col_span)
+    if row < 1 then return false end
+    if col < 1 then return false end
+    row_span = row_span > 0 and row_span or 1
+    col_span = col_span > 0 and col_span or 1
+    for i=row, (row + row_span - 1) do
+        if type(matrix[i]) == "table" then  -- check that the row exists
+            for j=col, (col + col_span -1) do
+                matrix[i][j] = nil
+            end
+        end
+    end
+    return true
+end
+
+--- Find a widget in a widget_table, by matching the coordinates in the matrix
+-- using the  `row`:`col` coordinates, and the spans `row_span` and `col_span`
+-- @param widgets_table Table of the widgets present in the grid
+-- @param number row Row number for the top left corner of the widget
+-- @param number col Column number for the top left corner of the widget
+-- @tparam number row_span The number of rows the widget spans (default to 1)
+-- @tparam number col_span The number of columns the widget spans (default to 1)
+-- @treturn table of index of widget_table
+local function find_widgets_by_coordinates(widgets_table, row, col, orig_row_span, orig_col_span)
+    local row_span = (orig_row_span and orig_row_span > 0) and orig_row_span or 1
+    local col_span = (orig_col_span and orig_col_span > 0) and orig_col_span or 1
+    local ret = {}
+    for index, data in ipairs(widgets_table) do
+        -- If one rectangular widget is on left side of other
+        local test_horizontal = not (row > data.row + data.row_span - 1 or data.row > row + row_span - 1)
+        -- If one rectangular widget is above other
+        local test_vertical   = not (col > data.col + data.col_span - 1 or data.col > col + col_span - 1)
```

Comment:
These conditions are giving me a headache...
Since the only caller only wants to check "is there any overlap with existing widgets", how about looping over the matrix instead and checking if all the needed entries do not exist? That sounds a lot simpler to me.